### PR TITLE
Document design decisions using ADRs - Versioning strategy ADR

### PR DESCRIPTION
**Pull request by jackgoodby**
# Description

Adds ADRs as a means of formally capturing design decisions for opg-data and specifically a Versioning strategy ADR

# Issues Resolved

IN-22

### CODE REVIEW

File: docs/adr/0002-api-versioning-strategy.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+If the version can't go in the URI or the body, that leaves headers.
+```
+GET /person/123 HTTP/1.1
+ApiVersion: 1.0
+Vary: ApiVersion
+```
+**Notable users:**
+* Azure
+
+Without the `Vary` header, it's hard for cache systems like Varnish to know that anything is different between two requests for the same resource at two different versions, and you could be served the wrong response, completely defeating the point.
```
Comment (mattmachell) [inline docs/adr/0002-api-versioning-strategy.md]: Most caches and proxies should be aware of vary. Certainly an issue once we get to 3rd party integrations, less of an issue with our internal ones.

File: docs/adr/0002-api-versioning-strategy.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+**Pros**
+* Quite RESTful
+
+**Cons**
+* non-obvious for developers, who are forced to consult docs
```
Comment (mattmachell) [inline docs/adr/0002-api-versioning-strategy.md]: Not sure of the phrasing here. Maybe "a less common pattern which might require more explanation for developers"

File: docs/adr/0002-api-versioning-strategy.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+Accept: application/json
+
+becomes
+
+Accept: application/vnd.opg-data.[version][+json]
+```
+
+**Pros**
+* Simple for API consumers (if they know about headers)
+* HATEOAS-friendly
```
Comment (mattmachell) [inline docs/adr/0002-api-versioning-strategy.md]: Worth referencing definition here.

File: docs/adr/0002-api-versioning-strategy.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+* Cache-friendly
+* RESTfully Keeps a URI for a resource the same
+* It opens up the possibility of adding resource-specific versioning in the future, eg `application/vnd.opg-data.donor.v1.1+json` would be a JSON representation of a donor resource at version 1.1
```
Comment (mattmachell) [inline docs/adr/0002-api-versioning-strategy.md]: If we go with semver, then we should really only be sending `application/vnd.opg-data.donor.v1+json` as only major versions should have breaking changes.

File: docs/adr/0002-api-versioning-strategy.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+**Cons**
+* API developers must be made aware about how the Accept header is being used
+* Harder to test - have to configure the accept header appropriately
```
Comment (mattmachell) [inline docs/adr/0002-api-versioning-strategy.md]: Is it a test thing though? Most HTTP clients and test frameworks give header control. Maybe more on the side of frameworks often don't implement header-based routing?

File: .adr-dir [hunk 6]
```diff
@@ -0,0 +1 @@
+docs/adr
```
Comment (mattmachell) [inline .adr-dir]: Also worth going from adr to architecture/decisions

File: docs/adr/0002-api-versioning-strategy.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+## Context
+
+The OPG-Data API must implement versioning. Must be RESTful, and offer as much flexibility with regard to future hosting/infrastructure as possible.
+
+There is no real consensus as there are several methods of versioning used by major players
+
+###TL;DR 
+####OPG-data Proposed Versioning Strategy (borrows the best bits from around the web)
+* We will implement versioning via the Content Negotiation using the Accept header, as per option 5, below. This seems the most future proof, most RESTful solution.
+* We will also support requests with a standard `Accept: application/json` - these will default to the latest version of the API (or should it be the oldest)
```
[the remark this answers is not in the document]
Comment (jackgoodby) [reply docs/adr/0002-api-versioning-strategy.md]: Incorporated suggestion

File: docs/adr/0002-api-versioning-strategy.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+* All breaking changes will be bundled into a new version
+* A Versions Timeline Document will be published and kept up to date with all breaking changes
```
[the remark this answers is not in the document]
Comment (jackgoodby) [reply docs/adr/0002-api-versioning-strategy.md]: Incorporated suggestion by stating we'll maintain a api/release-info endpoint

File: docs/adr/0002-api-versioning-strategy.md (hunk 1 above)
Comment (jackgoodby) [reply docs/adr/0002-api-versioning-strategy.md]: Couldn't see anything to do here - we're not plumping for this option anyway

File: docs/adr/0002-api-versioning-strategy.md (hunk 3 above)
Comment (jackgoodby) [reply docs/adr/0002-api-versioning-strategy.md]: I've written about HATEOAS and will cross ref when published

File: docs/adr/0002-api-versioning-strategy.md (hunk 4 above)
Comment (jackgoodby) [reply docs/adr/0002-api-versioning-strategy.md]: Agreed, I've updated

File: docs/adr/0002-api-versioning-strategy.md (hunk 5 above)
Comment (jackgoodby) [reply docs/adr/0002-api-versioning-strategy.md]: removed

File: .adr-dir (hunk 6 above)
Comment (jackgoodby) [reply .adr-dir]: Yep - done

File: docs/adr/0002-api-versioning-strategy.md (hunk 2 above)
Comment (jackgoodby) [reply docs/adr/0002-api-versioning-strategy.md]: Incorporated suggestion

**Comment by damacus** [comment]
LGTM
But it looks like we need some markdown linting on this markdown as it's wrong in a bunch of places (but correct enough to render)

**Comment by damacus** [comment]
FWIW, I think the pipeline should be enforing that @jackgoodby but thank you!

**Pull request merged**